src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs

```diff
@@ -160,9 +160,36 @@ private static ServiceContext CreateServiceContext(IOptions<KestrelServerOptions
 
                 async Task OnBind(ListenOptions options, CancellationToken onBindCancellationToken)
                 {
+                    var hasHttp1 = options.Protocols.HasFlag(HttpProtocols.Http1);
+                    var hasHttp2 = options.Protocols.HasFlag(HttpProtocols.Http2);
+                    var hasHttp3 = options.Protocols.HasFlag(HttpProtocols.Http3);
+                    var hasTls = options.IsTls;
+
+                    // Filter out invalid combinations.
+
+                    if (!hasTls)
+                    {
+                        // Http/1 without TLS, no-op HTTP/2 and 3.
+                        if (hasHttp1)
+                        {
+                            hasHttp2 = false;
+                            hasHttp3 = false;
+                        }
+                        // Http/3 requires TLS. Note we only let it fall back to HTTP/1, not HTTP/2
+                        else if (hasHttp3)
+                        {
+                            throw new InvalidOperationException("HTTP/3 requires UseHttps.");
+                        }
+                    }
+
+                    // Quic isn't registered if it's not supported, throw if we can't fall back to 1 or 2
+                    if (hasHttp3 && _multiplexedTransportFactory is null && !(hasHttp1 || hasHttp2))
```

Comment:
Have you considered what happens if someone wants to register a custom `IMultiplexedTransportFactory`?